### ISSUE TITLE
Multilanguage custom forms?

### ISSUE THREAD

**Issue by priethor** [Mailtrain-org/mailtrain#546]
Until now, I didn't require multilingual forms and when sending multilanguage campaigns we just created one per language and had the list divided in segments.

Forms, however, can't be assigned to segments, so the only solution that comes to mind is to have different lists per language. On the other hand, we have more segments in the lists (e.g., subscriber type) and having multiple lists per language would mean segments would be fractioned (we would not be able to just "filter by subscriber type" and get all subscribers, we would need to do that once per language list).

**TL;DR; What is the general approach people are using to manage multilanguage campaigns?**

**Comment by bures** [comment]
I don't think we have any consistent support for multilingual campaigns yet. At this point, I would say one list per language. Any consistent proposal for handling multi-language lists/campaigns is welcome. Thanks.

**Comment by priethor** [comment]
Hi @bures 

We are now able to contribute improving Mailtrain, as you have seen with Charlo270398's recent PRs. Following #627, we are trying to enable more than one custom form per List, so that each subscriber can recieve notifications and access forms in their own language.

However, there are some decissions that need to be taken as per how to manage who is assigned to what Custom form. At first we thought of segments, but a subscriber can belong to more than one segment, so at the moment our best approach is to add a new internal column called "language" so that lists have a default custom form or a different form depending on the language value.

What do you think about that? Do you think there is a better approach or would it be ok to add this new column?

**Comment by bures** [comment]
The more I think about it, I think that tying them to segments is actually not a bad idea. Also note that the language of the system default subscription form can be set by query string `locale` - e.g. `locale=pt-BR`. 

So I would suggest the following approach:
- It is possible to associate a custom form with a list and with every segment. For segment, this is optional.
- We add autogenerated `cid` field to custom forms and make it possible to choose a form by a query string `form=<form cid>`
- For existing subscriber, the form is determined in the following order:
  - If query string `form=<form cid>` is given, this particular form is used if it has been assigned to the list or one of its segments.
  - If a segment exists that specifies a form and the subscriber is a member of the segment, the form associated with the segment is used. If more segments apply, we choose the first segment. To have ordering of segments, we add `order` field to a segment (similar to fields)
  - The form from the list is used.

For a new subscriber, we determine the form from the query string `form=<form cid>`. If there is no such query string, we take the form associated with the list.

What do you think? This should cover situations when language is concerned, but also will allow other cases - like custom branding based on field in the list.

**Comment by priethor** [comment]
I really like the custom branding idea, since when we send segment-based campaigns we miss the option of showing them a more segment-branded form. However, I'm not sure if this can be confusing to the subscriber, since the more you customize the form, the more you lose the focus of what/where you are actually unsubscribing to/from. Moreover, in that regard I have the feeling that email notifications should always be the same for a subscriber, no matter the segment(s) he/she belongs to, in order to not lose perspective. Do you think the custom form ordering would solve the "perspective overview" issue?

Another thing we've noticed that led us to believe in a fixed `language` field for subscribers is that the custom forms URL usually has a language parameter, for example when you go to manage your preferences. However, we tried to change static texts according to this value and wasn't possible right off the bat, and as most of the custom form texts are not static but list-dependant(*), we didn't go that way. Do know what would be the purpose of this parameter? Maybe it's just being added by some middleware?

* Also, campaign field names cannot be translated just with segments, I think we still need an i18n concept to some extent, maybe combined with your segment proposal.

**Comment by priethor** [comment]
Hi Bures,

We are giving this issue a fresh thought, as our need to do multilingual forms keeps growing, and I'm afraid I don't understand what you mean by this:
>Also note that the language of the system default subscription form can be set by query string locale - e.g. locale=pt-BR.

Does it affect at all the form the locale set in the URL?

As for the labels and list column names, what would be your approach to translate them? i18n files? That would work for standard columns, but not for custom fields.

**Comment by forgeOneHQ** [comment]
We are going to start with the development and testing of mailtrain v3 in the next weeks.

You are welcome to help us with the testing as soon as the first release candidate is available.
